src/core/ddsc/src/dds_write.c

```diff
@@ -513,22 +550,43 @@ dds_return_t dds_write_impl (dds_writer *wr, const void * data, dds_time_t tstam
   thread_state_awake (ts1, &wr->m_entity.m_domain->gv);
 
   // 3. Check availability of iceoryx and reader status
-
   bool iceoryx_available = wr->m_iox_pub != NULL;
+  //bool has_fixed_size_type = wr->m_topic->m_stype->fixed_size;
+  void *iox_chunk = NULL;
 
   if(iceoryx_available) {
     //note: whether the data was loaned cannot be determined in the non-iceoryx case currently
     if(!deregister_pub_loan(wr, data))
     {
-      void* chunk_data = create_iox_chunk(wr);
-      if(chunk_data) {
-        memcpy (chunk_data, data, wr->m_topic->m_stype->iox_size);
-        data = chunk_data;
+      size_t required_size = get_required_buffer_size(wr->m_topic, data);
+      iox_chunk = create_iox_chunk(wr, required_size);     
+
+      if(iox_chunk) {
+        if(!fill_iox_chunk(wr, data, iox_chunk, required_size)) {
+          // serialization failed
+          ret = DDS_RETCODE_BAD_PARAMETER;
+          goto finalize_write;
+        }
+        data = iox_chunk;
       } else {
         // we failed to obtain a chunk, iceoryx transport is thus not available
-        // we will use the network path instead
+        // we cannot use the network path is now since due to the locators
+        // readers on the same machine would not get the data
+        // TODO: proper fallback logic to network path?
+
         iceoryx_available = false;
-      }
+        // TODO: activating this causes test failures
```

Comment:
Found it, or at least somewhat:
- the case where it needs to send via Iceoryx and also via the network ends up in the block starting at (new) line 620 (commented here with "X") and never sets `iox_chunk`. That in turn means the chunk never gets published, and after 8 (if I counted correctly) cases, it no longer can get a chunk from Iceoryx because there are too many chunks outstanding.
- the test then succeeds because `delivery_data_via_iceoryx` returns `false` if `iox_chunk` is a null pointer, and so the data gets delivered locally via Cyclone's native short-circuiting for local readers.

NB: This is not new in this PR, and clearly the test in `core/ddsi/tests/iceoryx.c` is too weak because it doesn't detect the underlying problem directly.

Reply:
I disabled this check in the test for now but I am not happy with it. Maybe I can fix this differently somehow from the description.

Reply:
I now release the chunk if `deliver_data_via_iceoryx` fails but I still need to disable the check in the tests, otherwise it fails. It appears the problem is not fully understood. Unless the test uses the `dds_loan` API I do not see from the current code path where we may forget to either publish or release an iceoryx chunk.

@eboasson

Reply:
I think I fixed it. I also had to add more cleanup code in case of failures. Can probably be optimized somewhat, but the lack of RAII in C makes cleanup error-prone and cumbersome. Regardlesss, the problem was passing the iox_chunk in specific cases after `convert_serdata` (when it behaves as an identity). This function probably should not even exist but somehow it does. Unfortunately it makes the control flow even more convoluted.

Reply:
That weird function does exist for a good reason: multi-sertype (in particular multi-language) support in a single process. E.g., if you have a single process written partly with a writer using the C++ API and a reader for the same topic using the C API, you need to convert in some manner. This does that by getting the serialised representation from the original `serdata` & deserializing for the reader.

What suprises me is that it'd be involved in the first place in this test because there's only a single language involved. Anyway, I'll study it now.